File: Learning-material/01-apis/ejemplos/webhooks/receiver.py
import hmac, hashlib
from fastapi import FastAPI, Request, HTTPException

SECRET = b"clave_secreta_compartida"
app = FastAPI()

def firma_valida(cuerpo: bytes, firma: str) -> bool:
    esperada = hmac.new(SECRET, cuerpo, hashlib.sha256).hexdigest()
    return hmac.compare_digest(esperada, firma)

@app.post("/webhooks/pagos")
async def recibir(request: Request):
    cuerpo = await request.body()
    firma = request.headers.get("X-Firma", "")
    if not firma_valida(cuerpo, firma):        # regla 1: verificar la firma
        raise HTTPException(400, "firma invalida")
    evento = await request.json()
    print(f"[receiver] evento OK: {evento['tipo']} monto={evento['monto']}")
    return {"recibido": True}                   # regla 2: responder rapido 200

File: Learning-material/01-apis/ejemplos/webhooks/receiver.js
const express = require("express");
const crypto = require("crypto");
const app = express();
const SECRET = "clave_secreta_compartida";

// necesitamos el body crudo para verificar la firma
app.use(express.raw({ type: "application/json" }));

function firmaValida(cuerpo, firma) {
  const esperada = crypto.createHmac("sha256", SECRET).update(cuerpo).digest("hex");
  return firma.length === esperada.length &&
         crypto.timingSafeEqual(Buffer.from(esperada), Buffer.from(firma));
}

app.post("/webhooks/pagos", (req, res) => {
  const firma = req.header("X-Firma") || "";
  if (!firmaValida(req.body, firma)) return res.status(400).json({ detail: "firma invalida" });
  const evento = JSON.parse(req.body);
  console.log(`[receiver] evento OK: ${evento.tipo} monto=${evento.monto}`);
  res.json({ recibido: true });
});
app.listen(8006, () => console.log("receiver en :8006"));
